### ISSUE TITLE
npm start error: No available ports found  on windows wsl

### ISSUE THREAD

**Issue by jifffffy** [remotion-dev/remotion#4315]
Error: No available ports found when I run npm start.
# Bug Report 🐛

I follow the office doc to create a test project  as follows:
```
# install deps
sudo apt install -y \
  libnss3 \
  libdbus-1-3 \
  libatk1.0-0 \
  libasound2t64 \  # libasound2 not exist refs https://github.com/cypress-io/cypress-documentation/issues/5816
  libxrandr2 \
  libxkbcommon-dev \
  libxfixes3 \
  libxcomposite1 \
  libxdamage1 \
  libgbm-dev \
  libatk-bridge2.0-0
 # create a project
npx create-video@latest
```
but i got an error when I run `npm start`:
![image](https://github.com/user-attachments/assets/ad711160-6061-4412-951d-24c3c190993c)

other infomation:
os: ubuntu 24.04
remotion:  4.0.212

**Comment by JonnyBurger** [comment]
Are you able to bind to any port between 3000 and 3001 with other frameworks?
Can you create a Next or Vite project for example?

**Comment by jifffffy** [comment]
yes, I am sure no ohter process bind port 3000 and 3001, 
it is ok when I  run on the windows and mac. this is not the first my remotion project, but this is the first in the wsl. 
I reboot the computer. but I still get the error.
thanks very much your quickly replay!

**Comment by jifffffy** [comment]
in get-port.ts file using the host addrest 127.0.0.1 and 0.0.0.0. cause  that socket connect timeout
127.0.0.1 and 0.0.0.0 in wsl cannot access with any port. should use a real ip address.

**Comment by JonnyBurger** [comment]
Thanks for investigating - What is a real IP address then?
If localhost is not available?

**Comment by JonnyBurger** [comment]
Maybe we should copy the logic from Vite

https://github.com/vitejs/vite/blob/21ec1ce7f041efa5cd781924f7bc536ab406a197/packages/vite/src/node/utils.ts#L933

**Comment by jifffffy** [comment]
The real address can get in the wsl as follows:
```
ip route show | grep -i default | awk '{ print $3}'
```
so it is good passing as a flag to the cli.

**Comment by JonnyBurger** [comment]
I am having a hard time getting WSL2 to run on my Windows for some reason.

Is somebody able to improve the logic in get-port.ts? https://github.com/remotion-dev/remotion/blob/main/packages/renderer/src/get-port.ts

Must be able to test it on WSL2!
Ideally it should not spawn a child process, surely there is a better way?

## Bounty

💎 This issue has a bounty on it!

Read our contributing guidelines: 
- **How to contribute**: https://www.remotion.dev/docs/contributing/
- **Implementing a new feature**: https://www.remotion.dev/docs/contributing/feature
- **Implementing a new option**: https://www.remotion.dev/docs/contributing/option
- **Write documentation**: https://www.remotion.dev/docs/contributing/docs
- **Bounty information**: https://www.remotion.dev/docs/contributing/bounty

/bounty 150

**Comment by hunxjunedo** [comment]
@JonnyBurger can I get assigned?

**Comment by JonnyBurger** [comment]
## No more attempts

This issue has been assigned to @hunxjunedo 🎉  

Do not send pull requests for this issue, do not ask to be in the queue, do not bother the assignee.  
**We will not accept any PRs not from the assignee and not award any bounties.**

Follow the #bounties channel on the Remotion Discord to get notified of new bounties.@hunxjunedo There you go!

**Comment by hunxjunedo** [comment]
@JonnyBurger its working fine on my end, @jifffffy can you provide more details so I can reproduce ?

![Image](https://github.com/user-attachments/assets/965b254c-f2ad-4b91-83f1-12474c139b46)

**Comment by hunxjunedo** [comment]
@jifffffy can you share the contents of `/etc/wsl.conf` using nano  or cat ?

**Comment by jifffffy** [comment]
@hunxjunedo I'm terribly sorry for the late reply. I've already tested it and everything is normal. Thank you all.

**Comment by jifffffy** [comment]
@hunxjunedo I'm terribly sorry for the late reply. the problem exists.  I've created a brand-new environment for testing today.
I  install system deps first: 
```
apt install -y \
  libnss3 \
  libdbus-1-3 \
  libatk1.0-0 \
  libasound2t64 \
  libxrandr2 \
  libxkbcommon-dev \
  libxfixes3 \
  libxcomposite1 \
  libxdamage1 \
  libgbm-dev \
  libatk-bridge2.0-0
```
then  I create a test project with the code-hike template
```
npx create-video@latest 
```
then I ensure Headless Shell is available, It might be optional.
```
npx remotion browser ensure
```
last I upgrade remotion to the latest version
```
npx remotion upgrade
```
now the package.json shows:
```
  "dependencies": {
    "@code-hike/lighter": "0.9.4",
    "@remotion/cli": "4.0.234",
    "@remotion/google-fonts": "4.0.234",
    "@remotion/layout-utils": "4.0.234",
    "@remotion/studio": "4.0.234",
    "codehike": "1.0.4",
    "polished": "4.3.1",
    "react": "18.3.1",
    "react-dom": "18.3.1",
    "remotion": "4.0.234",
    "twoslash-cdn": "0.2.9",
    "zod": "3.22.3"
  },
```
but when I run `npm run dev`, the error exist.
![Image](https://github.com/user-attachments/assets/14868c3f-75d2-4aa6-b161-6ed37c4605a2)

this is  [my test project](https://github.com/jifffffy/test-remotion.git)
this is my wsl version
![Image](https://github.com/user-attachments/assets/9e678e35-947c-4173-bcb6-1428c679f029)
os  version
![Image](https://github.com/user-attachments/assets/b0ac3524-6696-4ce5-b681-0e8447ccc7da)

**Comment by jifffffy** [comment]
if hostsToTest changes to hostsToTest.slice(0,1)  will be ok as  I said before.
```
const getPort = async ({ from, to, hostsToTest, }) => {
    const ports = makeRange(from, to);
    for (const port of ports) {
        if ((await (0, exports.testPortAvailableOnMultipleHosts)({
            port,
            hosts: hostsToTest.slice(0,1),
        })) === 'available') {
            return port;
        }
    }
    throw new Error('No available ports found');
};
```
![Image](https://github.com/user-attachments/assets/aeb156e0-c71c-4aba-8840-1e4a5534e468)

**Comment by hunxjunedo** [comment]
hey @jifffffy , I've tried to reproduce the issue many ways, even with the steps you mentioned, but still everything works fine:
![Image](https://github.com/user-attachments/assets/b5668652-378e-45b5-b16e-1f06836e2932)
![Image](https://github.com/user-attachments/assets/8110d63b-5cf5-41e3-a6ae-d6996aa9a160)

now if you ask me, I'm solid sure that there's something wrong with your wsl installation, what made me think that is the screenshot you shared in which you invoke `wsl --version` and it responds; there is no official version of wsl that accepts such paramter, rather it should not recognize `--version`. Can you please look into it ? And ensure you install wsl from the official source, like thru Ubuntu wsl on MS Store.

**Comment by jifffffy** [comment]
Thank you very much for  your replay. I've found the soulation. the problem is my wsl config.
I had this .wslconfig file in my home dir:
```
[wsl2]
networkingMode=mirrored
```
I removed it and shutdown wsl. Now, everythind is ok. Thanks again.
PS:  [WSL commands](https://learn.microsoft.com/en-us/windows/wsl/basic-commands) has `--version` flag.

**Comment by yrom** [comment]
> Thank you very much for your replay. I've found the soulation. the problem is my wsl config. I had this .wslconfig file in my home dir:
> 
> ```
> [wsl2]
> networkingMode=mirrored
> ```
> 
> I removed it and shutdown wsl. Now, everythind is ok. Thanks again. PS: [WSL commands](https://learn.microsoft.com/en-us/windows/wsl/basic-commands) has `--version` flag.

I  have this same issue.
After digging. I think the it should be a network-adapter bug of WSL2 (https://github.com/microsoft/WSL/issues/10855). 
The port availability detection of package `renderer` using `socket.connect()` which hangs for 3 seconds per port and always returns unavailable, even for ports that are clearly free.
See `isPortAvailableOnHost`  https://github.com/remotion-dev/remotion/blob/93c6f8f4b96c6c3b8963461e95c3b0e120b488a1/packages/renderer/src/get-port.ts#L6C1-L38C3

I think we can use `server.listen()` instead of `socket.connect()` for port detection:

```ts
const server = new net.Server();
server.once('error', () => resolve('unavailable'));
server.once('listening', () => {
    server.close();
    resolve('available');
});
server.listen(port, host); // Immediate result
```

**Comment by PatrickR-Binogi** [comment]
**Fix for end users that works right now:** Config.setStudioPort(42000) in remotion.config.ts — picks a port outside the Windows exclusion zone. Fragile since the exclusion ranges can shift between reboots. We confirmed this affects default WSL2 networking (not just mirrored mode)

We hit this exact issue on WSL2 without networkingMode=mirrored — just the default NAT mode (kernel 6.6.87.2-microsoft-standard-WSL2, Remotion 4.0.441).

Did a deep root cause analysis and can confirm @yrom's diagnosis.

Windows' WSL2 network layer accepts TCP connections on 127.0.0.1 and 0.0.0.0 across a large port range (~1024–41000),   even when nothing inside WSL listens there. This is part of WSL2's port forwarding proxy between the VM and the Windows host. Remotion's isPortAvailableOnHost uses socket.connect() to test availability — the WSL2 proxy accepts the connection, so the port appears "occupied."

We verified this by testing Remotion's exact availability check logic against all four hosts it uses:

  Port 3000:  ::1 → available, 127.0.0.1 → unavailable, :: → available, 0.0.0.0 → unavailable
  Port 8080:  same pattern
  Port 40000: same pattern
  Port 41000: all available 
  Port 42000: all available 
  Port 45000: 127.0.0.1 → unavailable again

Since testPortAvailableOnMultipleHosts requires ALL hosts to report available, every port in the 3000–3100 scan range fails. The studio HTTP server never starts — only webpack's HMR event stream server appears on a random high port, which gives the misleading impression that the studio launched.

**Proper fix:** Agree with @yrom that server.listen() is the right approach. A successful listen() proves the port is actually bindable,   rather than inferring availability from a failed connect(). This would fix it for all WSL2 configurations.
